recipes/LibriSpeech/SSL/train.py

```diff
@@ -0,0 +1,398 @@
+import sys
+import time
+import tempfile
+import pathlib
+import functools
+from contextlib import ExitStack
+
+from types import SimpleNamespace
+import speechbrain as sb
+import torchaudio
+import torch
+import torch.nn.functional as F
+from torch.nn.parallel import DistributedDataParallel
+from torch.utils.data import DataLoader
+from hyperpyyaml import load_hyperpyyaml
+
+from speechbrain import Stage
+from speechbrain.utils.distributed import run_on_main
+from speechbrain.dataio.dataloader import SaveableDataLoader, LoopedLoader
+from speechbrain.dataio.sampler import DynamicBatchSampler, DistributedSamplerWrapper
+from speechbrain.lobes.models.wav2vec import make_padding_mask, compute_mask, W2VPaddedBatch
+
+
+PYTHON_VERSION_MAJOR = 3
+PYTHON_VERSION_MINOR = 7
+
+
+logger = logging.getLogger(__name__)
+
+
+@sb.utils.checkpoints.register_checkpoint_hooks
+class W2V2Brain(sb.core.Brain):
+    def compute_forward(self, batch, stage):
+        """Computes forward pass through wav2vec model and returns encoded and
+        target embeddings as well as other metrics of interest.
+        """
+        batch = batch.to(self.device)
+        wavs, wav_lens, padding_mask, mask = batch.sig
+        wavs, wav_lens = wavs.to(self.device), wav_lens.to(self.device)
+        padding_mask, mask = padding_mask.to(self.device), mask.to(self.device)
+
+        # normalisation already done in dataloader
+        results = self.modules.model(wavs, wav_lens, padding_mask, mask, is_pretraining=True,
+            normalize_signal=False, do_final_projection=True)
+        return results
+
+    def compute_objectives(self, predictions, batch, stage):
+        """Samples negatives, computes contrastive loss and accuracy.
+        """
+        feats_targ = predictions["targets"]
+        embeddings = predictions["embeddings"]
+
+        negs = sample_negatives(feats_targ, 100)
+
+        loss, accuracy = self.hparams.loss(embeddings, feats_targ, negs)
+
+        objectives = {
+            "loss": loss,
+            "accuracy": accuracy,
+            "num_masked": predictions["num_masked"],
+            "ratio_masked": predictions["ratio_masked"],
+            "latent_l2": predictions["latent_l2"]
+        }
+        if "diversity_loss" in predictions:  # only quantised model has these
+            objectives.update(
+                {
+                    "diversity_loss": predictions["diversity_loss"],
+                    "prob_perplex": predictions["prob_perplex"],
+                    "code_perplex": predictions["code_perplex"],
+                    "num_vars": predictions["num_vars"],
+                    "temp": predictions["temp"],
+                }
+            )
+        return objectives
+
+    def fit_batch(self, batch):
+        should_step = self.step % self.grad_accumulation_factor == 0
+        # Managing automatic mixed precision
+        if self.auto_mix_prec:
+            with self.modules.model.no_sync() if not should_step else ExitStack():
```

Comment:
This I don't understand, what is the matter ?

Reply:
When doing DDP the gradients are synced on the backward pass, but when doing gradient accumulation every K steps then the syncing is only necessary every K steps and it's wasteful to do it on every backward pass. dan povey pointed it out in this PR https://github.com/speechbrain/speechbrain/pull/1251

the with .. if .. else ExitStack() is syntax so that the `with` statement is only executed when the condition is true